src/Compilers/CSharp/Portable/FlowAnalysis/PreciseAbstractFlowPass_Switch.cs

```diff
@@ -128,178 +128,44 @@ private void VisitSwitchSectionLabel(LabelSymbol label, BoundSwitchSection node)
-                        }
-
-                        _pendingBranches.Add(new PendingBranch(label, this.State));
-                    }
-                }
-            }
-
-            // we always consider the default label reachable for flow analysis purposes
-            // unless there was a single case that would match every input.
-            if (node.DefaultLabel != null)
-            {
-                if (node.SomeLabelAlwaysMatches)
-                {
-                    SetUnreachable();
-                }
-                else
-                {
-                    SetState(initialState.Clone());
-                }
-
-                _pendingBranches.Add(new PendingBranch(node.DefaultLabel, this.State));
-            }
-
-            // visit switch sections
-            for (var iSection = 0; iSection <= iLastSection; iSection++)
-            {
-                VisitPatternSwitchSection(switchSections[iSection], node.Expression, iSection == iLastSection);
-                // Even though it is illegal for the end of a switch section to be reachable, in erroneous
-                // code it may be reachable.  We treat that as an implicit break (branch to afterSwitchState).
-                IntersectWith(ref afterSwitchState, ref this.State);
-            }
-
-            SetState(afterSwitchState);
-        }
-
-        /// <summary>
-        /// Visit the switch expression, and return the initial break state.
-        /// </summary>
-        private LocalState VisitPatternSwitchHeader(BoundPatternSwitchStatement node)
-        {
-            // visit switch expression
-            VisitRvalue(node.Expression);
-
-            // return the exit state to use if no pattern matches
-            if (FullyHandlesItsInput(node))
-            {
-                return UnreachableState();
-            }
-            else
-            {
-                return this.State;
-            }
-        }
-
-        private bool FullyHandlesItsInput(BoundPatternSwitchStatement node)
-        {
-            // If the switch is complete (for example, because it has a default label),
-            // just return true.
-            if (node.IsComplete)
-            {
-                return true;
-            }
-
-            // We also check for completeness based on value. Other cases were handled in the construction of the decision tree.
-            if (node.Expression.ConstantValue == null)
-            {
-                return false;
-            }
-
+            var initialState = State.Clone();
+            HashSet<LabelSymbol> reachableLabels = node.DecisionDag.ReachableLabels;
             foreach (var section in node.SwitchSections)
             {
                 foreach (var label in section.SwitchLabels)
                 {
-                    if (label.Guard != null && label.Guard.ConstantValue != ConstantValue.True)
+                    if (reachableLabels.Contains(label.Label) || label.HasErrors)
```

Comment:
Didn't we just mark subsumed labels as erroneous in CheckSwitchErrors to avoid this check?

Reply:
No. Unreachable cases arise because of a constant switch governing expression. When the governing expression is constant, we replace the decision dag with a different one that has been reduced due to the constant expression, but after using the full decision dag to mark some as erroneous.

---
In reply to: [176226373](https://github.com/dotnet/roslyn/pull/24059#discussion_r176226373) [](ancestors = 176226373)